### PR TITLE
Fixing bug in MatrixElement printers

### LINKED ISSUE #11821: bug with MatrixSymbol and subs()
This session illustrates the bug:

In [1]: import sympy
In [2]: sympy.__version__
Out[2]: '1.0'
In [3]: from sympy import *
In [4]: A=MatrixSymbol('A',1,3)
In [5]: B=MatrixSymbol('B',1,3)
In [6]: C=MatrixSymbol('C',1,3)
In [7]: E=A-B
In [8]: F=C[0,0]
In [9]: F.subs(C,E)
Out[9]: (-1)*B + A[0, 0]

The output should be 
Out[9]: (-1)*B[0, 0] + A[0, 0]
or
Out[9]: ((-1)*B + A)[0, 0]

### PR DESCRIPTION
**Pull request by Alsheh**
I fixed #11821 all printing instances of the MatrixElement class and added test cases for each one of them.

### CODE REVIEW

**Comment by Alsheh** [comment]
@Upabjojr please review and it's not clear to me why there are conflicting files. I'd appreciate it if you could explain to me what the issue is. 

Thanks

File: sympy/printing/tests/test_ccode.py [hunk 1]
```diff
@@ -516,3 +516,13 @@ def test_ccode_For():
...
+
+def test_MatrixElement_printing():
+    # test case from issue #11821
+    A = MatrixSymbol("A",1,3)
+    B = MatrixSymbol("B",1,3)
+    C = MatrixSymbol("C",1,3)
+    E = A-B
+    F = C[0,0]
+    F = F.subs(C,E)
+    assert(ccode(F) == "((-1)*B + A)[0]")
```
Comment (Upabjojr) [inline sympy/printing/tests/test_ccode.py]: I think this is not valid in C.

File: sympy/printing/ccode.py [hunk 2]
```diff
@@ -219,8 +219,9 @@ def _print_ITE(self, expr):
         return self._print(_piecewise)
 
     def _print_MatrixElement(self, expr):
-        return "{0}[{1}]".format(expr.parent, expr.j +
-                expr.i*expr.parent.shape[1])
+        from sympy.printing.precedence import PRECEDENCE
+        return "{0}[{1}]".format(self.parenthesize(expr.parent,PRECEDENCE["Add"]),
```
Comment (Upabjojr) [inline sympy/printing/ccode.py]: Sorry, I was wrong about the precedence, it's not `Add`, it's `Atom`.

Anyway, I'm not sure about the C code. Did you check if the programs coming out actually do compile?

You cannot add arrays in C.

File: sympy/printing/ccode.py [hunk 3]
```diff
@@ -219,8 +219,9 @@ def _print_ITE(self, expr):
         return self._print(_piecewise)
 
     def _print_MatrixElement(self, expr):
-        return "{0}[{1}]".format(expr.parent, expr.j +
-                expr.i*expr.parent.shape[1])
+        from sympy.printing.precedence import PRECEDENCE
```
Comment (Upabjojr) [inline sympy/printing/ccode.py]: I believe this should be an import at the beginning of the file, not inside a function. Same for all other files.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 4]
```diff
@@ -5742,6 +5742,22 @@ def test_pretty_Mod():
...
 
+def test_MatrixElement_printing():
+    # test case from issue #11821
+
+    ascii_str1 = "(-B + A)[0, 0]"
+    ucode_str1 = u("(-B + A)[0, 0]")
+
+    A = MatrixSymbol("A",1,3)
+    B = MatrixSymbol("B",1,3)
+    C = MatrixSymbol("C",1,3)
```
Comment (Upabjojr) [inline sympy/printing/pretty/tests/test_pretty.py]: As a minor detail, put a space after the comma, we are trying to model the code after the PEP-8 recommendations.

File: sympy/printing/tests/test_ccode.py (hunk 1 above)
Comment (Alsheh) [inline sympy/printing/tests/test_ccode.py]: I'll double check that.

File: sympy/printing/ccode.py (hunk 3 above)
Comment (Alsheh) [inline sympy/printing/ccode.py]: Ok, I'll put imports at the top of the files.

Comment (Upabjojr) [inline sympy/printing/ccode.py]: My suggestion to new contributors is always to use PyCharm as Python editor, it can handle imports automatically for example.

File: sympy/printing/pretty/tests/test_pretty.py (hunk 4 above)
Comment (Alsheh) [inline sympy/printing/pretty/tests/test_pretty.py]: Will do.

File: sympy/printing/ccode.py (hunk 2 above)
Comment (Alsheh) [inline sympy/printing/ccode.py]: I will the precedence to Atom. See my comment regarding the C code above.

**Comment by Alsheh** [comment]
@Upabjojr , please review the new changes I made. 

Thanks.

File: sympy/printing/ccode.py [hunk 5]
```diff
@@ -219,9 +219,36 @@ def _print_ITE(self, expr):
...
+        from sympy.matrices import MatrixSymbol, MatrixExpr
+
+        if isinstance(expr.parent, MatrixSymbol):
+            return "{0}[{1}]".format(expr.parent, expr.j +
                 expr.i*expr.parent.shape[1])
 
+        terms = (expr.args[0].args)
+        PREC = precedence(expr.args[0])
+        l = []
+        for term in terms:
```
Comment (Upabjojr) [inline sympy/printing/ccode.py]: I think this code is better off in the `MatrixElement` class. We may want to rewrite:
```python
>>> ((x*A+y*B)[0, 1])._some_expansion_method()
x*A[0, 1] + y*B[0, 1]
```
That is, rewrite the expression by distributing the element access.

I'd suggest to leave this for another PR, so let's (temporarily) ignore the C code printer. If you want, better open a separate PR, this code should be separately tested.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 6]
```diff
@@ -5742,6 +5742,48 @@ def test_pretty_Mod():
...
+    assert upretty(F) == ucode_str1
+
+    ascii_str1 = "(A + M)[0, 1]"
+    ucode_str1 = u("(A + M)[0, 1]")
+    E = A + M
+    F = C[0, 1]
+    F = F.subs(C, E)
+    assert pretty(F)  == ascii_str1
+    assert upretty(F) == ucode_str1
+
```
Comment (Alsheh) [inline sympy/printing/pretty/tests/test_pretty.py]: >>> E = 2*x*A + 3*Matrix([[x, y, z]])
>>> F = C[0, 0]
>>> F = F.subs(C, E)
([3⋅x  3⋅y  3⋅z] + 2⋅x⋅A)[0, 0]

This is now a test case.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 7]
```diff
@@ -5742,6 +5742,48 @@ def test_pretty_Mod():
...
+    ucode_str1 = u("(-B + A)[0, 0]")
+    E = A-B
+    F = C[0, 0]
+    F = F.subs(C, E)
+    assert pretty(F)  == ascii_str1
+    assert upretty(F) == ucode_str1
+
+    ascii_str1 = "(-B + A + M)[0, 0]"
+    ucode_str1 = u("(-B + A + M)[0, 0]")
+    E = A - B + M
```
Comment (Alsheh) [inline sympy/printing/pretty/tests/test_pretty.py]: Additional test cases were added.

File: sympy/printing/tests/test_latex.py [hunk 8]
```diff
@@ -1586,6 +1587,31 @@ def test_issue_10489():
...
+    E = A - B + M
+    F = C[0, 0]
+    F = F.subs(C, E)
+    assert latex(F) == r"\left(-1 B + A + M\right)_{0, 0}"
+
+    E = A + M
+    F = C[0, 1]
+    F = F.subs(C, E)
+    assert latex(F) == r"\left(A + M\right)_{0, 1}"
+
```
Comment (Alsheh) [inline sympy/printing/tests/test_latex.py]: Those test cases were added.

File: sympy/printing/ccode.py [hunk 9]
```diff
@@ -219,9 +219,36 @@ def _print_ITE(self, expr):
...
+            t = self._print(term)
+            if t.startswith('-'):
```
Comment (Alsheh) [inline sympy/printing/ccode.py]: I am just trying to keep track of the signs here. I plan to correct the `(-1)*B` bug in a separate PR.

File: sympy/printing/tests/test_fcode.py [hunk 10]
```diff
@@ -675,3 +675,28 @@ def test_fcode_For():
...
+
+    E = A-B
+    F = C[0, 0]
+    F = F.subs(C, E)
+    assert(fcode(F) == "      ((-1)*B + A)(1, 1)")
+
+    E = A - B + M
+    F = C[0, 0]
+    F = F.subs(C, E)
+    assert(fcode(F) == "      ((-1)*B + A + M)(1, 1)")
```
Comment (Alsheh) [inline sympy/printing/tests/test_fcode.py]: @asmeurer @moorepants @smichr or @jksuom review this please.

**Comment by asmeurer** [comment]
Looks like there are merge conflicts (probably because of your other PR that I just merged).

**Comment by Alsheh** [comment]
@asmeurer I just fixed the merge conflict.

File: sympy/printing/ccode.py (hunk 9 above)
Comment (Upabjojr) [inline sympy/printing/ccode.py]: I still believe that this code is in the wrong location. It should be moved to `_print_MatAdd` and re-adapted there. I wouldn't merge this PR with this code in this position.

I also fear this code could potentially introduce bugs, as it might potentially be unable to deal with some special cases of matrix elements.

So, either remove this code from this PR and put it in a new PR, or fix `_print_MatAdd` in this PR.

Comment (Alsheh) [inline sympy/printing/ccode.py]: @Upabjojr please see my second attempt to fix this issue here https://github.com/sympy/sympy/pull/11917

**Comment by Upabjojr** [comment]
@Alsheh let's finish this PR before starting another one. This is practically ready to merge, except for the extra code that I wouldn't merge.

**Comment by gxyd** [comment]
@Upabjojr does this add anything more compared to @smichr 's PR https://github.com/sympy/sympy/pull/12800 ?
Though this bug:
```python
In [4]: B = MatrixSymbol('B',1,3)
In [9]: B*(-1)
Out[9]: (-1)*B

In [10]: -B
Out[10]: (-1)*B
```
Remains to be fixed. Probably we can close this PR, since this also doesn't fix this bug either.

**Comment by czgdp1807** [comment]
Closing in reference to https://github.com/sympy/sympy/pull/11839#issuecomment-355771282 and https://github.com/sympy/sympy/issues/11821#issuecomment-548232668
Thanks for your contributions.
